doc/UltiSnips.txt

```diff
@@ -302,7 +303,33 @@ then you define your mapping as >
 and if the you can't expand or jump from the current location then the
 alternative function IMAP_Jumpfunc('', 0) is called.
 
-3.2.2 Path to Python module                   *UltiSnips-python-module-path*
+3.2.2 Inner mappings                          *UltiSnips-inner-mappings*
+--------------------
+
+                                *UltiSnipsMapInnerKeys* *UltiSnipsUnmapInnerKeys*
+For maximum compatibility with other plug-ins, UltiSnips sets up some special
+mappings when a snippet starts being expanded, and tears them down once
+expansion is completed. In order to make it possible to override these inner
+mappings, it fires the following "User" autocommands:
+
+UltiSnipsMapInnerKeys
+UltiSnipsUnmapInnerKeys
+
+For example, to call a pair of custom functions in response to these events,
+you might do: >
+
+   autocmd! User UltiSnipsMapInnerKeys
+   autocmd User UltiSnipsMapInnerKeys call CustomInnerKeyMapper()
+   autocmd! User UltiSnipsUnmapInnerKeys
+   autocmd User UltiSnipsUnmapInnerKeys call CustomInnerKeyUnmapper()
+
+Note that because snippet expansion may be nested, it is possible that
```

Comment:
Am I misreading the implementation? `_setup_inner_state` and `_teardown_inner_state` look asymmetrical to me, which implies the names should be `UltiSnipsEnterSnippet` (not `EnterFirst`) and `UltiSnipsExitLastSnippet`).

`_setup_inner_state` gets called repeatedly (and runs) when nesting snippets, but `_teardown_inner_state` has a guard (via `_current_snippet_is_done`) that prevents it from being called until the last snippet it exited.

(See the updated tests that I just pushed to https://github.com/wincent/ultisnips/commits/autocommands-squashed that show that we `_setup_inner_state` gets called twice for a snippet nested inside a snippet, but `_teardown_inner_state` only gets called once.)

It would be nice to have symmetrical names, but then they'd be not totally accurate because the methods aren't actually symmetrical.